external/qjs_crypto_module.c

```diff
@@ -0,0 +1,678 @@
+
+
+typedef struct {
+    njs_str_t      name;
+
+    size_t         size;
+    njs_hash_init  init;
+    njs_hash_update update;
+    njs_hash_final final;
+} njs_hash_alg_t;
+
+typedef struct {
+    JSContext      *cx;
+    njs_hash_t     ctx;
+    njs_hash_alg_t *alg;
+} njs_digest_t;
+
+typedef struct {
+    u_char         opad[64];
+    JSContext      *cx;
+    njs_hash_t     ctx;
+    njs_hash_alg_t *alg;
+} njs_hmac_t;
+
+
+typedef struct {
+    njs_str_t          name;
+
+    qjs_digest_encode  encode;
+} qjs_crypto_enc_t;
+
+
+
+static njs_hash_alg_t * qjs_crypto_algorithm(JSContext *cx, JSValueConst val);
+static qjs_crypto_enc_t * qjs_crypto_encoding(JSContext *cx, JSValueConst val);
+static void qjs_hash_finalizer(JSRuntime *rt, JSValue val);
+static void qjs_hmac_finalizer(JSRuntime *rt, JSValue val);
+static JSValue qjs_crypto_create_hash(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv);
+static JS_BOOL qjs_is_typed_array(JSContext *cx, JSValue val);
+static JSValue qjs_hash_prototype_update(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv, int hmac);
+static JSValue qjs_hash_prototype_digest(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv, int hmac);
+static JSValue qjs_hash_prototype_copy(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv);
+static JSValue qjs_crypto_create_hmac(JSContext *cx, JSValueConst this_val,
+    int argc, JSValueConst *argv);
+static int qjs_crypto_module_init(JSContext *cx, JSModuleDef *m);
+static JSModuleDef * qjs_crypto_init(JSContext *cx, const char *module_name);
+
+
+static JSValue
+qjs_buffer_digest(JSContext *cx, const njs_str_t *src)
+{
+    return qjs_buffer_create(cx, src->start, src->length);
+}
+
+
+static njs_hash_alg_t njs_hash_algorithms[] = {
+
+    {
+        njs_str("md5"),
+        16,
+        njs_md5_init,
+        njs_md5_update,
+        njs_md5_final
+    },
+
+    {
+        njs_str("sha1"),
+        20,
+        njs_sha1_init,
+        njs_sha1_update,
+        njs_sha1_final
+    },
+
+    {
+        njs_str("sha256"),
+        32,
+        njs_sha2_init,
+        njs_sha2_update,
+        njs_sha2_final
+    },
+
+    {
+        njs_null_str,
+        0,
+        NULL,
+        NULL,
+        NULL
+    }
+
+};
+
+
+static qjs_crypto_enc_t njs_encodings[] = {
+
+    {
+        njs_str("buffer"),
+        qjs_buffer_digest
+    },
+
+    {
+        njs_str("hex"),
+        qjs_string_hex
+    },
+
+    {
+        njs_str("base64"),
+        qjs_string_base64
+    },
+
+    {
+        njs_str("base64url"),
+        qjs_string_base64url
+    },
+
+    {
+        njs_null_str,
+        NULL
+    }
+
+};
+
+
+static njs_hash_alg_t *
+qjs_crypto_algorithm(JSContext *cx, JSValueConst val)
+{
+    njs_str_t       name;
+    const char      *alg_str;
+    njs_hash_alg_t  *a, *alg;
+
+    alg_str  = JS_ToCString(cx, val);
+    if (njs_slow_path(alg_str == NULL)) {
```

Comment:
Avoid using `njs_slow_path()` and `njs_fast_path()`. We are moving away from them in the new code. Here and below.